settings/rust-analyzer.vim

```diff
@@ -159,6 +162,47 @@ function! s:on_find_matching_brace(x) abort
     endif
 endfunction
 
+function! s:expand_macro() abort
+    echo 'Expanding Macros'
+    call lsp#callbag#pipe(
+        \  lsp#request('rust-analyzer', {
+        \   'method': 'rust-analyzer/expandMacro',
+        \   'params': {
+        \       'textDocument': lsp#get_text_document_identifier(),
+        \       'position': lsp#get_position(),
+        \   },
+        \ }),
+        \ lsp#callbag#subscribe({
+        \   'next': {x->s:on_expand_macro(x)},
+        \   'error': {e->lsp_settings#utils#error(e)},
+        \ })
+        \ )
+endfunction
+
+function! s:on_expand_macro(x) abort
+    if lsp#client#is_error(a:x['response']) | echom 'Failed to expand macro' | endif
+    let l:contents = a:x['response']['result']['expansion']
+
+    let l:lines = lsp#utils#_split_by_eol(l:contents)
+    let l:view = winsaveview()
+    let l:alternate=@#
+    silent! pclose
+    sp LspRustExpandMacro
+    execute 'resize '.min([len(l:lines), &previewheight])
+    set previewwindow
+    setlocal conceallevel=2
+    setlocal bufhidden=hide
+    setlocal nobuflisted
+    setlocal buftype=nofile
+    setlocal ft=rust
```

Comment:
Hi, Thank you for great work!
I've try this ane love it. But I have some feed backs.

I use `g:lsp_diagnostics_virtual_text_enabled = true`.
Execute `:LspRustExpandMacro` and move to preview window.
Shows a lot of errors and bit a annoying 😰 

https://github.com/user-attachments/assets/efd87f7b-5cb3-4b1e-8b27-ccb6e7847cc4

I'm not sure my suggestion is right or not, but set no filetype and enable syntax for preview looks good to me.

```diff
diff --git a/settings/rust-analyzer.vim b/settings/rust-analyzer.vim
index b71d6eb..dde8fd7 100644
--- a/settings/rust-analyzer.vim
+++ b/settings/rust-analyzer.vim
@@ -194,10 +194,13 @@ function! s:on_expand_macro(x) abort
     setlocal bufhidden=hide
     setlocal nobuflisted
     setlocal buftype=nofile
-    setlocal ft=rust
     setlocal noswapfile
     %d
     call setline(1, l:lines)
+
+    let l:win_id = win_getid()
+    call win_execute(l:win_id, $'setlocal syntax=rust')
+
     execute "normal \<c-w>p"
     call winrestview(l:view)
     let @#=l:alternate
```

https://github.com/user-attachments/assets/fa25d652-0d89-4430-b955-0fc4e4690c5f

@mattn Do you have any good solutions?

Reply:
Oh nice! I don't have the virtual diagnostics enabled but definitely see how it's rough. I think the syntax set to rust works. I've added those changes.